File: Other Parsers/parserFromClassWithInput.c
/* parser.c - A syntax analyzer system for simple arithmetic expressions */

#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>

/* Global Variables */
int charClass;
char lexeme[100];
char nextChar;
int lexLen;
int nextToken;
char inputString[256];  // Stores user input
int currentIndex = 0;   // Tracks current position in input

/* Function Declarations */
void addChar();
void getChar();
void getNonBlank();
int lex();
void expr();
void term();
void factor();
void error(const char *message);

/* Character Classes */
#define LETTER 0
#define DIGIT 1
#define UNKNOWN 99

/* Token Codes */
#define INT_LIT 10
#define IDENT 11
#define ASSIGN_OP 20
#define ADD_OP 21
#define SUB_OP 22
#define MULT_OP 23
#define DIV_OP 24
#define LEFT_PAREN 25
#define RIGHT_PAREN 26
#define EOF_TOKEN -1

/******************************************************/
/* Main Driver */
int main() {
    /* Get user input */
    printf("Enter an arithmetic expression: ");
    fgets(inputString, sizeof(inputString), stdin);

    /* Remove newline character if present */
    size_t len = strlen(inputString);
    if (len > 0 && inputString[len - 1] == '\n') {
        inputString[len - 1] = '\0';
    }

    /* Start processing the input */
    currentIndex = 0;
    getChar();
    lex();
    expr();

    /* Ensure the input is fully consumed */
    if (nextToken != EOF_TOKEN) {
        error("Unexpected tokens after expression");
    }

    printf("Parsing complete!\n");
    return 0;
}

/*****************************************************/
/* Look up operators and parentheses */
int lookup(char ch) {
    switch (ch) {
        case '(':
            addChar();
            nextToken = LEFT_PAREN;
            break;
        case ')':
            addChar();
            nextToken = RIGHT_PAREN;
            break;
        case '+':
            addChar();
            nextToken = ADD_OP;
            break;
        case '-':
            addChar();
            nextToken = SUB_OP;
            break;
        case '*':
            addChar();
            nextToken = MULT_OP;
            break;
        case '/':
            addChar();
            nextToken = DIV_OP;
            break;
        default:
            addChar();
            nextToken = EOF_TOKEN;
            break;
    }
    return nextToken;
}

/*****************************************************/
/* Add nextChar to lexeme */
void addChar() {
    if (lexLen <= 98) {
        lexeme[lexLen++] = nextChar;
        lexeme[lexLen] = '\0';
    } else {
        printf("Error - lexeme is too long\n");
    }
}

/*****************************************************/
/* Get the next character from user input */
void getChar() {
    if (currentIndex < strlen(inputString)) {
        nextChar = inputString[currentIndex++];
        if (isalpha(nextChar))
            charClass = LETTER;
        else if (isdigit(nextChar))
            charClass = DIGIT;
        else
            charClass = UNKNOWN;
    } else {
        charClass = EOF_TOKEN;
        nextChar = '\0';
    }
}

/*****************************************************/
/* Skip whitespace characters */
void getNonBlank() {
    while (isspace(nextChar))
        getChar();
}

/*****************************************************/
/* Lexical Analyzer */
int lex() {
    lexLen = 0;
    getNonBlank();

    switch (charClass) {
        case LETTER:
            addChar();
            getChar();
            while (charClass == LETTER || charClass == DIGIT) {
                addChar();
                getChar();
            }
            nextToken = IDENT;
            break;

        case DIGIT:
            addChar();
            getChar();
            while (charClass == DIGIT) {
                addChar();
                getChar();
            }
            nextToken = INT_LIT;
            break;

        case UNKNOWN:
            lookup(nextChar);
            getChar();
            break;

        case EOF_TOKEN:
            nextToken = EOF_TOKEN;
            strcpy(lexeme, "EOF");
            break;
    }

    printf("Next token is: %d Next lexeme is %s\n", nextToken, lexeme);
    return nextToken;
}

/*****************************************************/
/* expr: Parses <expr> -> <term> {(+ | -) <term>} */
void expr() {
    printf("Enter <expr>\n");
    term();
    while (nextToken == ADD_OP || nextToken == SUB_OP) {
        lex();
        term();
    }
    printf("Exit <expr>\n");
}

/*****************************************************/
/* term: Parses <term> -> <factor> {(* | /) <factor>} */
void term() {
    printf("Enter <term>\n");
    factor();
    while (nextToken == MULT_OP || nextToken == DIV_OP) {
        lex();
        factor();
    }
    printf("Exit <term>\n");
}

/*****************************************************/
/* factor: Parses <factor> -> id | int_constant | ( <expr> ) */
void factor() {
    printf("Enter <factor>\n");

    if (nextToken == IDENT || nextToken == INT_LIT) {
        lex();
    } else if (nextToken == LEFT_PAREN) {
        lex();
        expr();
        if (nextToken == RIGHT_PAREN) {
            lex();
        } else {
            error("factor() - Missing right parenthesis");
        }
    } else {
        error("factor() - Invalid token");
    }

    printf("Exit <factor>\n");
}

/*****************************************************/
/* Error handling function */
void error(const char *message) {
    printf("Syntax Error: %s\n", message);
    exit(1);
}

File: Other Parsers/parserFromClassWithInput.py
import sys

# Global variables
input_string = ""  # User input string
current_index = 0  # Current position in the input string
next_token = None  # Current token
next_char = ""  # Current character
char_class = None  # Class of current character
lexeme = ""  # Stores lexeme

# Character classes
LETTER = 0
DIGIT = 1
UNKNOWN = 99

# Token codes
INT_LIT = 10
IDENT = 11
ADD_OP = 21
SUB_OP = 22
MULT_OP = 23
DIV_OP = 24
LEFT_PAREN = 25
RIGHT_PAREN = 26
EOF_TOKEN = -1 # END OF FILE - Really interesting to understand why this is so important

# Side note: the EOF indicates the end of an input. An example for this is when on C storing a character string has always at the end the null empty char '\0'. Really interesting to see that it comes from an actual grammar rule.

# Function Declarations:

def lookup(ch):
    """
    Lookup table of operators that returns the appropiate token based on next operator
    """
    global next_token # To access the next token
    token_dict = {
        '(': LEFT_PAREN,
        ')': RIGHT_PAREN,
        '+': ADD_OP,
        '-': SUB_OP,
        '*': MULT_OP,
        '/': DIV_OP
    }
    next_token = token_dict.get(ch, EOF_TOKEN)
    return next_token


def get_char():
    """
    Reads the next char and determines its class (letter, digit, unk, or EOF).
    """
    global next_char, char_class, current_index # Retrieves next character, its class and index

    if current_index < len(input_string):
        next_char = input_string[current_index]
        current_index += 1

        if next_char.isalpha():
            char_class = LETTER
        elif next_char.isdigit():
            char_class = DIGIT
        else:
            char_class = UNKNOWN
    else:
        char_class = EOF_TOKEN
        next_char = ''  # Empty string represents EOF


def get_non_blank():
    """
    Skips whitespaces.
    """
    while next_char.isspace():
        get_char()


def add_char():
    """
    Adds the current character to the lexeme.
    """
    global lexeme # retrieves current lexeme
    lexeme += next_char


def lex():
    """
    Lexical analyzer to tokenize the input.
    """
    global lexeme, next_token
    lexeme = ""
    get_non_blank()

    if char_class == LETTER:
        #  IDs/Variable Names
        add_char()
        get_char()
        while char_class in (LETTER, DIGIT):
            add_char()
            get_char()
        next_token = IDENT

    elif char_class == DIGIT:
        # Integer
        add_char()
        get_char()
        while char_class == DIGIT:
            add_char()
            get_char()
        next_token = INT_LIT

    elif char_class == UNKNOWN:
        # Parentheses/Operators
        lookup(next_char)
        add_char()
        get_char()

    elif char_class == EOF_TOKEN:
        next_token = EOF_TOKEN
        lexeme = "EOF"

    print(f"Next token is: {next_token} Next lexeme is {lexeme}")
    return next_token


def error(location):
    """
    Error validation function.
    """
    print(f"Syntax Error in {location}")
    sys.exit(1)

# Grammar section of the program
# This was the easiest to translate, it's literally just translating the CFG

def expr():
    """
    
    <expr> ::= <term> { ("+" | "-") <term> }
    
    """
    print("Enter <expr>")
    term()
    while next_token in (ADD_OP, SUB_OP):
        lex()
        term()
    print("Exit <expr>")


def term():
    """
    <term> ::= <factor> { ("*" | "/") <factor> }
    """
    print("Enter <term>")
    factor()
    while next_token in (MULT_OP, DIV_OP):
        lex()
        factor()
    print("Exit <term>")


def factor():
    """
    <factor> ::= <IDENT> | <INT_LIT> | "(" <expr> ")" 
    
    """
    print("Enter <factor>")

    if next_token in (IDENT, INT_LIT):
        lex()
    elif next_token == LEFT_PAREN:
        lex()
        expr()
        if next_token == RIGHT_PAREN:
            lex()
        else:
            error("factor() - Missing right parenthesis")
    else:
        error("factor() - Invalid token")

    print("Exit <factor>")


def main():
    global input_string, current_index

    # Get user input
    input_string = input("Enter the arithmetic expression: ") + " "  # Add space to handle EOF detection
    current_index = 0

    # Lexical analysis
    get_char()
    lex()

    # Parser
    expr()

    # Handling EOF errors
    if next_token != EOF_TOKEN:
        error("main() - Unexpected tokens after expression")

    print("Parsing complete!")

if __name__ == "__main__":
    main()
